File: 2023/python/day01.py
from pathlib import Path

from loguru import logger
from typer import Typer
from utils import timer

main = Typer()


def calibration(line: str) -> int:
    """Compute calibration

    Identify first and larg integer in the string, multiplying the first by 10
    and summing them.
    """
    if len(line.strip()) == 0:
        return 0

    first: int | None = None
    last: int = -1
    for c in line:
        if c.isdigit():
            first = first or int(c)
            last = int(c)

    return (first or 0) * 10 + last


spelling = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
]


def replace_letter_numbers(line: str) -> str:
    """Replace written numbers by digits

    oneight is interpreted as 18
    """
    res = []
    for offset in range(len(line)):
        if line[offset].isdigit():
            res.append(line[offset])
            continue

        for i, number in enumerate(spelling):
            if line[offset:].startswith(number):
                res.append(str(i + 1))

    return "".join(res)


@timer
def task01(input: str) -> int:
    """Task 01 solution"""
    return sum(calibration(line) for line in input.split("\n"))


@timer
def task02(input: str) -> int:
    """Task 02 solution"""
    return sum(calibration(replace_letter_numbers(line)) for line in input.split("\n"))


@main.command()
def entrypoint(path: Path):
    """Entrypoint reading the input

    Args:
        path: Path to input file
    """
    with open(path) as f:
        input = f.read()

    logger.info("Task 01")
    logger.info(task01(input))

    logger.info("Task 02")
    logger.info(task02(input))


if __name__ == "__main__":
    main()

File: 2023/go/day01.go
package main

import (
	"fmt"
	"os"
	"strings"
	"unicode"

	"github.com/urfave/cli/v2"
)

func calibration(line string) int {
	first := 0
	last := 0

	for _, c := range line {
		if unicode.IsDigit(c) {
			last = int(c - '0')
			if first == 0 {
				first = last
			}
		}
	}
	return 10 * first + last
}

func replaceLetterNumbers(line string) string {
	numbers := [...]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
	var res []rune
	for i, c := range line {
		if unicode.IsDigit(c) {
			res = append(res, c)
		} else {
			for j, number := range numbers {
				if strings.HasPrefix(line[i:], number) {
					res = append(res, rune(j + 1) + '0')
				}
			}
		}
	}

	return string(res)
}

func task01(input string) int {
	res := 0
	for _, line := range strings.Split(input, "\n") {
		res += calibration(line)
	}

	return res
}

func task02(input string) int {
	res := 0
	for _, line := range strings.Split(input, "\n") {
		res += calibration(replaceLetterNumbers(line))
	}

	return res
}

func entrypoint(c *cli.Context) error {
	path := c.Args().First()
	input, err := os.ReadFile(path)

	if err != nil {
		panic(err)
	}

	fmt.Printf("Task 01: %d\n", task01(string(input)))
	fmt.Printf("Task 02: %d\n", task02(string(input)))

	return nil
}

func main() {
	app := &cli.App{
		Name: "day01",
		Description: "Advent of Code 2023 - Day 01",
		Action: entrypoint,
	}

	if err := app.Run(os.Args); err != nil {
		panic(err)
	}
}
